Feat: Feature gene annotation

grantn5 opened the pull request:
I have written some code that will produce a per gene CNV matrix, giving the mean log2 CNV per gene (as 1 gene can appear in many windows) 
I have also updated the `np.convolve` mode to `valid` so it matches the documentation in the function. Full methods [here](https://numpy.org/doc/stable/reference/generated/numpy.convolve.html)

I have also updated the code in the instance where the window is larger than the number of genes on the chromosome, to only run one convolution the length of the genes on the given chr see https://github.com/grantn5/infercnvpy/blob/feature_geneAnnotation/src/infercnvpy/tl/_infercnv.py#L198

I have also updated the `genomic_position_from_gtf` function so it will always read the gtf in as a pandas df (avoids bug where it reads it in as polars df) 

```
gtf = gtfparse.read_gtf(
        gtf_file, usecols=["seqname", "feature", "start", "end", "gene_id", "gene_name"], result_type="pandas"
    )
```

Closes #128

grantn5 commented:
Hi @grst I am not familiar with this building of docs. It is failing due to a warning because the sphinx can't seem to find scanpy in the build environment, any suggestions?

grst commented:
It seems the `pca` function got moved within scanpy. I fixed it in another PR. 
Once you resolved the merge conflicts, readthedocs should build again.

grst commented on src/infercnvpy/io/_genepos.py:
```diff
@@ -39,8 +39,8 @@ def genomic_position_from_gtf(
         If True, add the annotations directly to adata, otherwise return a dataframe.
     """
     gtf = gtfparse.read_gtf(
-        gtf_file, usecols=["seqname", "feature", "start", "end", "gene_id", "gene_name"]
-    ).to_pandas()
```
What exactly was the bug you were facing here? Shouldn't the new and the old version be equivalent? 

Just asking because I remember some back-and-forth on this line in the past, because some versions of gtfparse are incompatible with some polars version.

grantn5 replied:
Hi @grst, I realise that the PR I was working on was a previous version of inferCNVpy that did not have the `.to_pandas`, therefore I was getting an error relating to indexes. However you are correct, setting `.to_pandas` and `result_type= pandas` are equivalent. I will revert

grst commented on src/infercnvpy/tl/__init__.py:
```diff
@@ -41,24 +41,24 @@ def pca(
 ) -> Union[np.ndarray, None]:
     """Compute the PCA on the result of :func:`infercnvpy.tl.infercnv`.
 
-    Thin wrapper around :func:`scanpy.tl.pca`.
```
The correct reference is now `scanpy.pp.pca` which is why intersphinx was failing.

grantn5 commented:
Hi @grst It looks like all tests that use the oligodendroglioma adata are failing? 
Has this object been moved in scanpy?

grst commented:
I think this is rather an incompatibility with numpy 2.0 which was very recently released. 
I'll look into it, but the issue might be in one of the dependencies.

grst commented on src/infercnvpy/tl/_infercnv.py:
```diff
@@ -161,26 +181,83 @@ def _running_mean(x: np.ndarray | scipy.sparse.spmatrix, n: int = 50, step: int
-def _running_mean_by_chromosome(expr, var, window_size, step) -> tuple[dict, np.ndarray]:
+        return smoothed_x, convolved_gene_values
+
+    else:  # If there is less genes than the window size, set the window size to the number of genes and perform a single convolution
+        n = x.shape[1]  # set the filter size to the number of genes
+        r = np.arange(1, n + 1)
+        ## As we are only doing one convolution the values should be equal
+        pyramid = np.array([1] * n)
+        # Old code to make the pyramid
+        # pyramid = np.minimum(r, r[::-1])
+        smoothed_x = np.apply_along_axis(
+            lambda row: np.convolve(row, pyramid, mode="valid"),
+            axis=1,
+            arr=x,
+        ) / np.sum(pyramid)
+
+        ## As all genes are used the convolution the values are identical for all genes
+        convolved_gene_values = pd.DataFrame(np.repeat(smoothed_x, len(gene_list), axis=1), columns=gene_list)
+
+        return smoothed_x, convolved_gene_values
```
Now the if and else block are even more similar than before. Do you think we could get rid of the code duplication?

grst commented on src/infercnvpy/tl/_infercnv.py:
```diff
@@ -291,17 +378,20 @@ def _infercnv_chunk(tmp_x, var, reference, lfc_cap, window_size, step, dynamic_t
+        gene_res[np.abs(gene_res) < noise_thres] = 0
 
     x_res = scipy.sparse.csr_matrix(x_res)
+    # Reindex column, filling unused columns with NA
+    # gene_res = gene_res.reindex(columns=var.index, fill_value=np.nan)
+    # gene_res = scipy.sparse.csr_matrix(gene_res)
```
Seems this can be removed?

grst reviewed (commented):
Just some minor comments. Thanks for adding test cases :) 

Have you compared the runtime before and after adding your code? It looks like it could make the whole workflow significantly slower and memory intense. In that case it would be great if it could be disabled via a function parameter in `tl.infercnvpy`.

grst commented on src/infercnvpy/tl/_infercnv.py:
```diff
@@ -161,26 +181,83 @@ def _running_mean(x: np.ndarray | scipy.sparse.spmatrix, n: int = 50, step: int
-        smoothed_x = smoothed_x / pyramid.sum()
-        return smoothed_x
+        ## get the indices of the genes used in the convolution
+        convolution_indices = get_convolution_indices(x, n)[np.arange(0, smoothed_x.shape[1], step)]
```
Can the `step` be integrated into get_convolution_indices directly? Currently, a large array of indices is generated, but only a fraction of it is ever used (because of `step`).

grantn5 commented:
> Just some minor comments. Thanks for adding test cases :)
> 
> Have you compared the runtime before and after adding your code? It looks like it could make the whole workflow significantly slower and memory intense. In that case it would be great if it could be disabled via a function parameter in `tl.infercnvpy`.

Hi thanks for the review!

I have tested the time on our in-house data ~50,000 cells, ~20,000 genes and it runs within about 10 minutes parallelized over 32 cpu's (before it took ~2 mins). However due to the nested for loop you have to massively reduce the chunksize ~100 as each sample needs to be looped through the flattened index of genes for each window https://github.com/grantn5/infercnvpy/blob/feature_geneAnnotation/src/infercnvpy/tl/_infercnv.py#L230 which will be very long.

I also found that there is a bug in n_jobs where it isn't automatically parallelized over all available cpus's so I have always been setting it manually.  

This is definitely a CPU bound task rather than memory so if you are using a machine with lots of CPUs it will speed up immensely. 

If required for the PR to be accepted I can do further refactoring to add a parameter to turn off per gene CNV calling by default. However, this may take some time due to the way the function is handling outputs from intermediate functions.

grantn5 commented on src/infercnvpy/tl/_infercnv.py:
```diff
@@ -123,14 +124,27 @@ def infercnv(
+    # that are not included in the CNV will be filled with NaN
+    per_gene_df = per_gene_df.reindex(columns=adata.var_names, fill_value=np.nan)
+    # This needs to be a sparse matrix otherwise the memory usage is too high
+    per_gene_mtx = scipy.sparse.csr_matrix(per_gene_df)
```
Hi sorry this is a poorly worded comment, you are correct the matrix is already dense so converting it is unnecessary. My rationale for this is the column names. When it was kept as a data frame I ran into an error with saving the h5ad object with `adata.write` that was something along the lines of `colnames are too large to be saved compressed into h5ad`. Hence the converstion to sparse as it does not have colnames. 

This is also why this https://github.com/grantn5/infercnvpy/blob/feature_geneAnnotation/src/infercnvpy/tl/_infercnv.py#L137 line of code is required so the adata.var_names matches the colnames of the per_gene_df, so you can map back the values for each gene.

grst replied:
Ok, but then you could just convert the data frame to a dense numpy array using
```
per_gene_mtrx = per_gene_df.values
```

grst commented:
Thanks for testing this. Given that it's a 5x runtime increase and scalability is one of the main selling points of this library, I would like to be able to switch this off. For 50k cells it might not be much of an issue, but when working with atlases >1M cells, this makes quite a difference.

grantn5 commented:
Hi @grst I have made the calculation of per gene CNV optional, default set to `False`, I have also added some documentation to the function. 

Additionally, I have added some benchmarking to the pytests to show the performance of the added computation. 
Hopefully this is now good merge!

grst reviewed (approved):
LGTM, thanks!

knadia07 commented:
hi there, i am confused how do we actually get the output for feature gene annotation mapping to back chromosome position, say for cells with very high cnv score?
